Mathlib/Tactic/TFAE.lean

```diff
@@ -0,0 +1,224 @@
+/-
+Copyright (c) 2018 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin, Reid Barton, Simon Hudon, Thomas Murrills
+-/
+import Lean
+import Mathlib.Tactic.Have
+import Mathlib.Tactic.SolveByElim
+import Mathlib.Data.List.TFAE
+import Qq.Match
+
+/-!
+# The Following Are Equivalent (TFAE)
+
+This file provides the tactics `tfae_have` and `tfae_finish` for proving goals of the form
+`TFAE [P₁, P₂, ...]`.
+-/
+
+open List Lean Meta Expr Elab.Term Elab.Tactic Mathlib.Tactic Qq
+
+namespace Mathlib.Tactic.TFAE
+
+/-- An arrow of the form `←`, `→`, or `↔`. -/
+syntax impArrow := " → " <|> " ↔ " <|> " ← "
+
+/--
+`tfae_have` introduces hypotheses for proving goals of the form `TFAE [P₁, P₂, ...]`. Specifically,
+`tfae_have : i arrow j` introduces a hypothesis of type `Pᵢ arrow Pⱼ` to the local context,
+where `arrow` can be `→`, `←`, or `↔`. Note that `i` and `j` are natural number indices (beginning
+at 1) used to specify the propositions `P₁, P₂, ...` that appear in the `TFAE` goal list. A proof
+is required afterward, typically via a tactic block.
+
+```lean
+example (h : P → R) : TFAE [P, Q, R] := by
+  tfae_have : 1 → 3
+  { exact h }
+  ...
+```
+The resulting context now includes `tfae_1_to_3 : P → R`.
+
+The introduced hypothesis can be given a custom name, in analogy to `have` syntax:
+```lean
+tfae_have h : 2 ↔ 3
+```
+
+Once sufficient hypotheses have been introduced by `tfae_have`, `tfae_finish` can be used to close
+the goal.
+
+```lean
+example : TFAE [P, Q, R] := by
+  tfae_have : 1 → 2
+  { /- proof of P → Q -/ }
+  tfae_have : 2 → 1
+  { /- proof of Q → P -/ }
+  tfae_have : 2 ↔ 3
+  { /- proof of Q ↔ R -/ }
+  tfae_finish
+```
+-/
+syntax (name := tfaeHave) "tfae_have " (ident " : ")? num impArrow num : tactic
```

Comment:
I'm cargo-culting, but should this be:
```suggestion
syntax (name := tfaeHave) "tfae_have " (binderIdent " : ")? num impArrow num : tactic
```

Reply:
(Also, sorry for the delayed response; I fell prey to GitHub's "pending" label and thought I had taken all the action I needed to. 🙃)

![image](https://user-images.githubusercontent.com/68410468/218339075-cc1c7a88-d21a-4f4b-8587-ec9f6e55c996.jpeg)

Reply:
`impArrow` question resolved [on zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/refactoring.20mathport.20syntax.3F).